### ISSUE TITLE
Toplevel's lexer drop the content of its buffer after reading a sentence

### ISSUE THREAD

**Issue by jhjourdan** [ocaml/ocaml#13486]
Let's do an experiment:

```
$ cat > test.ml
Printf.printf "a\n";; Printf.printf "b\n";;
Printf.printf "c\n";;
$ ocaml test.ml  # Interpreter mode
a
b
c
$ ocaml < test.ml  # Toplevel mode
OCaml version 5.2.0
Enter #help;; for help.

# a
- : unit = ()
b
- : unit = ()
# c
- : unit = ()
# 
```

That is, the end of the first line of the input is just ignored. My interpretation is that the lexer drops the content of its buffer when it sees the first occurrence of `;;`, and since the toplevel reads input line by line, the buffer only contains the code up to the end of the line.

This is confirmed by increasing the length of the first line: at some point, some suffix of the first line is lexed and parsed., because not all the line fits in the buffer.

**Comment by gasche** [comment]
This should have been fixed by #12029, so from a distance I don't understand why we have the problem again.

**Comment by stedolan** [comment]
I don't understand the bug here: OCaml 5.2.0 is printing `a`, `b` and `c` as instructed. What's wrong in this example? (To me it looks like #12029 is working as intended?)

**Comment by gasche** [comment]
I agree -- sorry I missed this earlier. Let's close optimistically. (We can always reopen if @jhjourdan explains that there is a bug.)

**Comment by jhjourdan** [comment]
Sorry for the confusion. I can reproduce the bug on 5.1.1, but not on 5.3.0, so indeed #12029 probably fixed the bug.
